mivideo/mivideo_extract.py

```diff
@@ -0,0 +1,134 @@
+# -*- coding: utf-8 -*-
+'''
+Module for setting up and running the MiVideo data extract.
+'''
+
+import logging
+import os
+import time
+from typing import Union, Sequence, Dict
+
+import pandas as pd
+from google.cloud import bigquery
+from google.oauth2 import service_account
+from sqlalchemy.engine import ResultProxy
+
+import mivideo.queries as queries
+from db.db_creator import DBCreator
+from environ import ENV, CONFIG_DIR
+from vocab import ValidDataSourceName
+
+logger = logging.getLogger(__name__)
+
+SHAPE_ROWS: int = 0  # Index of row count in DataFrame.shape() array
+
+
+class MiVideoExtract(object):
+    '''
+    Initialize the MiVideo data extract process by instantiating the ``MiVideoExtract`` class,
+    then invoke its ``run()`` method.
+
+    For example, ``MiVideoExtract().run()``.
+    '''
+
+    def __init__(self):
+        self.udpKeyFilePath: str = os.path.join(CONFIG_DIR, ENV.get('mivideo', {}).get(
+            'service_account_json_filename'))
+        logger.debug(f'udpKeyFilePath: "{self.udpKeyFilePath}"')
+
+        self.credentials: service_account.Credentials = \
+            service_account.Credentials.from_service_account_file(
+                self.udpKeyFilePath,
+                scopes=['https://www.googleapis.com/auth/cloud-platform'],
+            )
+
+        self.udpDb: bigquery.Client = bigquery.Client(
+            credentials=self.credentials,
+            project=self.credentials.project_id,
+        )
+
+        logger.info(f'Connected to BigQuery project: "{self.udpDb.project}"')
+
+        DB_PARAMS = ENV['INVENTORY_DB']
+        APPEND_TABLE_NAMES = ENV.get('APPEND_TABLE_NAMES', [
+            'job_run', 'data_source_status', 'mivideo_media_started_hourly',
+            'mivideo_media_creation'])
+
+        self.appDb: DBCreator = DBCreator(DB_PARAMS, APPEND_TABLE_NAMES)
+        self.appDb.set_up()
```

Comment:
If you're just going to use `engine`, you shouldn't need to call `set_up`.

Reply:
Oh?  I'll see…

Reply:
It's needed.  When I commented out the `set_up()` call, I saw messages like this in the log:

```txt
course_inventory | DEBUG:google.api_core.retry:Retrying due to , sleeping 16.7s ...
```

It was repeated several times, with various random values in place of "16.7s".

When I uncommented the line, the code worked again.

The reason I thought I needed to call `set_up()` is because it calls `engine.connect()`.  So, we could either call that in the MiVideo code, or we could call `set_up()`.  (0.5 = ½)

Reply:
Okay, yeah, sorry for the confusion. I saw you using `result: ResultProxy = self.appDb.engine.execute(sql)`, and I kind of thought you were bypassing the connection setup part. I created `set_up` and `tear_down` to initialize and close the `self.conn` object, but you aren't using that, so I wasn't sure if it was necessary. If it were me, I'd be calling `self.appDB.conn.execute(sql)` and using `tear_down` afterward.

I am not sure why you are seeing `google` errors, this seems completely unrelated.

Reply:
Alternatively, you could use `pandas.read_sql` and `self.engine` for getting that last value, and not worry about `set_up` and `tear_down`.

Reply:
@lsloan, FWIW, I tried and removing the `set_up` and had no problem getting this to run.

Reply:
Good suggestions, but I think I'll leave it as it is for now.